### ISSUE TITLE
delayP

### ISSUE THREAD

**Issue by char0n** [char0n/ramda-adjunct#1099]
This function has two behaviors:

1. Creates a promise that resolves after x milliseconds. This function is overloaded and instead of `Number` can also consume `options` object. If timeout not specified by options object, the returned promise resolves immediately.
```js
RA.delayP(100); //=> resolves after 100 milliseconds with `undefined` value
RA.delayP({ timeout: 100, value: "val" }); //=> resolves after 100 milliseconds with `"val"` value
```

2. Create a promise that rejects after the specified milliseconds. This function is overloaded and instead of `Number` can also consume `options` object. If timeout not specified by options object, the returns promise rejects immediately.
```js
RA.delayP.reject(100); //=> rejects after 100 milliseconds with `undefined` value
RA.delayP.reject({ timeout: 100, value: new Error('error') }); //=> rejects after 100 milliseconds with `Error('error')` value
```

[This PR](https://github.com/char0n/ramda-adjunct/pull/1067) should serve as a model template of PR for `delayP` function.

**Comment by PranjalAgni** [comment]
Can I try this?

**Comment by char0n** [comment]
@PranjalAgni go ahead. It's yours!

**Comment by PranjalAgni** [comment]
I am new to functional programming. Can you help me about how I can add .reject to my delayP function?

**Comment by char0n** [comment]
No problem. Let's say you already defined a `delayP` function as following.
```js
const delayP = ...;
```
In javascript, even a functiona is an object so you can assign the `reject` function directly to `delayP` function in following way:
```js
const delayP = ...;

delayP.reject = (arg) => ...function definition;
```
And that's it. Let me know if the explanation was clear. Example how to do it can also be found here: https://github.com/char0n/ramda-adjunct/blob/b9ac42e0a182958f5930218b4a3bde900c7be4ea/src/lensIso.js#L48

**Comment by PranjalAgni** [comment]
Thanks for the explanation it was really easy to catch

**Comment by PranjalAgni** [comment]
Hey @char0n I was going through the codebase there are these two functions `resolveP` and `rejectP` so can I use them internally when writing `delayP`.
Also  can you quickly tell me why we are using `bind` everywhere?

**Comment by PranjalAgni** [comment]
Also when you wrap your functions with `curry1` how do you pass paramters to it. As in many functions I don't find where you have passed argument for it for e.g. `allEqual` fn.

**Comment by char0n** [comment]
> Hey @char0n I was going through the codebase there are these two functions resolveP and rejectP
> so can I use them internally when writing delayP.

Yeah, pls use them internaly.

> Also can you quickly tell me why we are using bind everywhere?

This library can be used in very old environments theoretically. In old JavaScript envs not all functions where static, and were bound to the context that they were called from. For example following code would not work in some javascript env.s
```js
const { log } = console;
log('message');
```
But the following would.
```js
const { log } = console;
log.call(console, 'message');
```
We're trying to write code in a way so that we increase the possibility to run on as much js envs as possible. That's why use use bind even if not necessary in new js envs.


> Also when you wrap your functions with curry1 how do you pass paramters to it. As in many 
> functions I don't find where you have passed argument for it for e.g. allEqual fn.

Can you rephrase this question ?
Curry1 works in a following way:

```js
const inc = curry1(number => number + 1);

inc(1) === 2;
```
`curry1` consumes a function and returns a function which accepts one paramter. This parameter is then passed to the function that you're wrapping with `curry1`

**Comment by PranjalAgni** [comment]
Yeah got you explained me how `curry1` works.

**Comment by PranjalAgni** [comment]
Also I am using `pipe` to call `resolveP` how can I  send the value param to `resolveP`?

**Comment by PranjalAgni** [comment]
> Also I am using `pipe` to call `resolveP` how can I send the value param to `resolveP`?

Figured it out!

**Comment by PranjalAgni** [comment]
Currently I am using `setTimeout` to make a delay and resolving a promise inside it  but then I need to use `.then` to get the resolved value from `delayP` for e.g `delayP().then(val => console.log(val))`. Should I use javascript `sleep()` so then I don't need to put a `.then` after `delayP()`

**Comment by char0n** [comment]
To figure it out what you're trying to do I'd need some code fragment.

**Comment by PranjalAgni** [comment]
Sure

**Comment by PranjalAgni** [comment]
The problem I am facing is I need to add a `.then()` to my `delayP()` to make it work. I am sharing a sample code which I was playing with:
```js
const makeDelay = time => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      return resolve('Yes');
    }, time);
  });
};


const delayP = time => {
  return makeDelay(time).then(val => {
    console.log(val); // Yes
    return val;
  });
};

console.log(delayP(1000)); // Promise {}

delayP(1000).then(val => {
  console.log(val); // Yes
});
```

I wanted a behaviour like:
```js
console.log(delayP(1000)) // Yes
```

**Comment by PranjalAgni** [comment]
Also is there a way in current setup to execute only single file in the test directory. If so please share the command for it.

**Comment by char0n** [comment]
> Also is there a way in current setup to execute only single file in the test directory. If so please share the command for it.

There is. By calling `.only`:

```js
describe.only("name", function() {
});
```
or
```js
describe.only("name", function() {
  context.only("name", function() {
    specify.only("name", function() {
    });
    it.only("name", function() {
    });
  });
});
```
Only the most inner block with `only` call will be called when running tests.

**Comment by char0n** [comment]
@PranjalAgni the behavior of `console.log(delayP(1000))` is not possible. The function must always return promise.

**Comment by char0n** [comment]
This example 
```js
delayP(1000).then(val => {
  console.log(val); // Yes
});
```
reflects the reality what we really want.

**Comment by PranjalAgni** [comment]
> > Also is there a way in current setup to execute only single file in the test directory. If so please share the command for it.
> 
> There is. By calling `.only`:
> 
> ```js
> describe.only("name", function() {
> });
> ```
> 
> or
> 
> ```js
> describe.only("name", function() {
>   context.only("name", function() {
>     specify.only("name", function() {
>     });
>     it.only("name", function() {
>     });
>   });
> });
> ```
> 
> Only the most inner block with `only` call will be called when running tests.

I was currently using `mocha` grep pattern for this `mocha --grep <NAME_OF_YOUR_TEST>`

**Comment by PranjalAgni** [comment]
> This example
> 
> ```js
> delayP(1000).then(val => {
>   console.log(val); // Yes
> });
> ```
> 
> reflects the reality what we really want.

I have implemented the function like this so it is good to go.
Also should I throw a error message if argument passed by the user is not a `Number` or `Object`

**Comment by char0n** [comment]
> I was currently using mocha grep pattern for this mocha --grep <NAME_OF_YOUR_TEST>

Yes you can also use the grep, there are multiple ways how to achieve the same.

> I have implemented the function like this so it is good to go.
Also should I throw a error message if argument passed by the user is not a Number or Object

Nope, we try not to validate input arguments unless absolutely necessary.
